### PR TITLE
[SPARK-33784][SQL] Rename dataSourceRewriteRules batch

### PR DESCRIPTION
**Pull request by aokolnychyi**
### What changes were proposed in this pull request?

This PR tries to rename `dataSourceRewriteRules` into something more generic.

### Why are the changes needed?

These changes are needed to address the post-review discussion [here](https://github.com/apache/spark/pull/30558#discussion_r533885837).

### Does this PR introduce _any_ user-facing change?

Yes but the changes haven't been released yet.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala
Comment (aokolnychyi) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: This isn't final.

I'll explain my current thinking here. We are going to use this batch to rewrite plans directly after operator optimization. I did not go for `planRewriteRules` due to the [comment](https://github.com/apache/spark/pull/30558#discussion_r533151935) made by @dongjoon-hyun and @viirya (please let me know if you don't feel that way anymore). I also did not go for `preCBORules` as there quite a few things that will happen before CBO like discussed in the original thread.

**Comment by aokolnychyi** [comment]
cc @rdblue @HyukjinKwon @dongjoon-hyun @cloud-fan @viirya @dbtsai @sunchao @gatorsmile @maryannxue

**Comment by dongjoon-hyun** [comment]
Thank you for making a follow-up, @aokolnychyi . I believe we can reach an agreement here.

**Comment by rdblue** [comment]
I'm fine with post operator optimization.

**Review by viirya — COMMENTED** [review]
Sounds better to me as the name indicates the position of the batch should be.

**Review by HyukjinKwon — APPROVED** [review]
This looks good to me. @gatorsmile @maryannxue @cloud-fan do you have any though?

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: To be honest, I do not like this name. Our catalyst is designed as a query compiler. There does not exist a concept in a query compiler called "Post Operator Optimization". If you talk about this extension point in a Database class, no one understands what it means from the name. 

My point is still the same. Here, we are adding the extension point for advanced users to add the rules between the rule-based optimizer (RBO) rules and the cost-based optimizer (CBO) rules.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: The batch/function/rule names in Optimizer.scala is not critical to me. They are private. We can change them whenever we need. However, this API is an external developer API. We should be very careful to name it. 

To me, either preCBORules or postRBORules are much better than postOperatorOptimizationRule.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: BTW, eventually, we should move all the statistics based rules from the optimizer to the planner

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: It's unfortunate that we don't have a clear separation between RBO and CBO. There are RBO rules before and after the only CBO rule `CostBasedJoinReorder`.

I think the general idea of this batch is to allow people to inject special optimizer rules that can't be run together with the main operator optimizer batch. It's indeed a Spark specific thing, as the main operator optimizer batch will be run many times until reaching the fixed point, the new batch added here will be run only once.

It's really hard to do the naming here. To match the actual purpose and to be general, how about `Phase 2 Optimizer Rules` or `Run Once Optimizer Rules`?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: `Phase 2 Optimizer Rules` or `Run Once Optimizer Rules` does not explain the location of the batch. 

How many batches do we want to add? I am afraid we will add phase 2, 3, 4, .... This is endless. It looks very random. We should not expect the users need to understand/read the source code of our optimizer every new release. It is also fragile.

**Comment by gatorsmile** [comment]
Let us mark as WIP before we get an agreement on the API.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala
Comment (aokolnychyi) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: I do share some of the points brought by @cloud-fan. There is no clear separation between CBO and RBO and we run quite some rule-based optimizations after the existing cost-based optimization rule. Also, we need to run this batch before early scan pushdown rules to capture possible rewrites, which, in turn, run before CBO. That’s why `preCBORules` does seem to capture what this rule does.

Our existing hook in the extensions API says `The injected rules will be executed during the operator optimization batch`. That was one of my motivations for the current name as we have resolution and post-hoc resolution rules.

That said, I don’t mean `postOperatorOptimizationRules` is a perfect name either. It just seems to better reflect the place in the Spark optimizer. I’ll be happy to discuss and iterate further.

Comment (aokolnychyi) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:215]: I'd be also interested to hear from @hvanhovell and other folks who commented.

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, All.

All of your comments are important to us and @aokolnychyi has been trying to embrace your opinions and to lead the discussion on the naming according to @gatorsmile 's comment. (https://github.com/apache/spark/pull/30558#discussion_r533885837)

> The batch name and comments look not good to me. We need a better name here.

Since this issue (SPARK-33784) is also considered as an blocker by @HyukjinKwon , let's try to summarize once more. The following is a brief summary. Please feel free to add.
| Name | Description |
|-|-|
| customRewriteRules | The initial naming from the author |
| customDataSourceRewriteRules | The initial commit with 5 `+1` |
| postOperatorOptimizationRules | This PR |
| preCBORules | 3 `+1` and 1 `-1` |
| planRewriteRules, preCostBasedOptimizationRules | [The other proposed alternatives](https://github.com/apache/spark/pull/30558#discussion_r543286779) |

AFAIK, according to the current comments, `preCBORules` was the most actively supported by three people as the better alternative and vetoed by one.
- @maryannxue proposed https://github.com/apache/spark/pull/30558#discussion_r543017895
- @cloud-fan wrote `I'd vote for preCBORules.`
- @gatorsmile wrote `To me, either preCBORules or postRBORules are much better than postOperatorOptimizationRule`
- @rdblue wrote `I don't think that preCBORules is a good name.`

However, the committed name, `customDataSourceRewriteRules`, also had +5 approval.

In terms of `veto` status, although there are unclear parts between the communication, the following two looks outstanding.
- @gatorsmile vetoed `dataSourceRewriteRules` after +5 approval
- @rdblue vetoed `preCBORules` after +3 positive comments.

Do you think you can switch your decision one way or another to make an agreement for Apache Spark 3.1.0, @gatorsmile and @rdblue ? Given the design of AS-IS Spark, I don't think we can find a perfect name in Spark 3.2 timeframe, either.

**Comment by rdblue** [comment]
I think that `preCBORules` is a bad name, but I am fine with using it to move forward.

**Comment by dongjoon-hyun** [comment]
Thank you for making us move forward, @rdblue .

@aokolnychyi , could you update this PR once more to get further comments?

**Comment by dongjoon-hyun** [comment]
Also, @gatorsmile . Please let us know what you think about the AS-IS direction.

**Comment by HyukjinKwon** [comment]
I am okay with `preCBORules` too. Looks like we're okay with this name.

**Comment by aokolnychyi** [comment]
I am going to update this PR to use `preCBORules` if there is enough consesus. I did have similar concerns to what @cloud-fan mentioned in [this](https://github.com/apache/spark/pull/30808#discussion_r544787602) comment but I don't have an ideal alternative.

Let me change the name and unblock this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala
Comment (aokolnychyi) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:211]: I added that the rule will be executed once to the docs.

**Comment by HyukjinKwon** [comment]
Seems we reached to the agreement. I am removing `WIP`.

**Comment by aokolnychyi** [comment]
I've updated this PR. 

Thanks everyone who participated in the discussion and @dongjoon-hyun for creating a summary to move this forward.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for update. I hope we can have this functionality in Apache Spark 3.1.0.

**Review by gatorsmile — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Comment by cloud-fan** [comment]
It conflicts with 3.1, @aokolnychyi  can you open a backport PR? thanks!

**Comment by aokolnychyi** [comment]
Yep, will do today, @cloud-fan.

**Comment by dongjoon-hyun** [comment]
Thank you all for the decision!

**Comment by dongjoon-hyun** [comment]
Gentle ping, @aokolnychyi .

**Comment by aokolnychyi** [comment]
@dongjoon-hyun @cloud-fan, shall I cherry-pick this change in full or shall I cherry-pick only part that is present in 3.1?

Like I mentioned earlier, the second part of this change (the hook to inject custom rules) was created before 3.1 was cut but was merged a couple of days after. I think it only makes sense to have the new batch in 3.1 only if we also add the hook.

**Comment by dongjoon-hyun** [comment]
Sorry for late response. The new backporting PR (SPARK-33621 and SPARK-33784) looks good to me and it landed at branch-3.1 for Apache Spark 3.1.0 a few minutes ago.

**Pull request closed without merging**
